sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/QueryStage.scala

```diff
@@ -178,64 +86,87 @@ abstract class QueryStage extends UnaryExecNode {
       prefix: String = "",
       addSuffix: Boolean = false,
       maxFields: Int): Unit = {
-    child.generateTreeString(
+    val planToDisplay = if (finalPlanReady) {
+      finalPlan
+    } else {
+      originalPlan
+    }
+    planToDisplay.generateTreeString(
       depth, lastChildren, append, verbose, "", false, maxFields)
   }
 }
 
+
 /**
  * The last QueryStage of an execution plan.
  */
-case class ResultQueryStage(var child: SparkPlan) extends QueryStage
+case class ResultQueryStage(originalPlan: SparkPlan) extends QueryStage {
+
+  override def start(): Future[Unit] = {
+    throw new IllegalStateException(
+      "ResultQueryStage is the root node and should be started directly in `execute`")
+  }
+
+  // Caches the created output RDD so we can reuse that.
+  private var cachedRDD: RDD[InternalRow] = null
+
+  override def doExecute(): RDD[InternalRow] = synchronized {
+    if (cachedRDD == null) {
+      cachedRDD = finalPlan.execute()
+      val executionId = sparkContext.getLocalProperty(SQLExecution.EXECUTION_ID_KEY)
+      // It's possible that we need to get the RDD without SQLExecution, e.g. `Dataset.rdd`.
+      if (executionId != null) {
+        sparkContext.listenerBus.post(SparkListenerSQLAdaptiveExecutionUpdate(
+          executionId.toLong,
+          SQLExecution.getQueryExecution(executionId.toLong).toString,
+          SparkPlanInfo.fromSparkPlan(finalPlan)))
+      }
+    }
+    cachedRDD
+  }
+
+  def planToRun: SparkPlan = finalPlan
```

Comment:
What's the thinking behind `planToRun`, maybe directly use finalPlan is clean semantics.